File: Scripts/ex8_2_1.m
% exercise 8.2.1

% Load data
cdir = fileparts(mfilename('fullpath')); 
load(fullfile(cdir,'../Data/synth5'))

%% Fit model using bootstrap aggregation (bagging)

% Number of rounds of bagging
L = 100;

% Variable for model parameters
w_est = nan(M+1, L);

% Weights for selecting samples in each bootstrap
weights = ones(N,1)/N;

% For each round of bagging
for l = 1:L
    % Choose data objects by random sampling with replacement 
    i = discreternd(weights, N);

    % Extract training set
    X_train = X(i, :);
    y_train = y(i);

    % Fit logistic regression model to training data and save result
    w_est(:,l) = glmfit(X_train, y_train, 'binomial');
end

% Evaluate the logistic regression on the training data
p = glmval(w_est, X, 'logit');

% Estimated value of class labels (using 0.5 as threshold) by majority voting
y_est = sum(p>.5,2)>L/2; 

% Compute error rate
ErrorRate = sum(y~=y_est)/N;
fprintf('Error rate: %.0f%%\n', ErrorRate*100);

%% Plot decision boundary
mfig('Decision boundary'); clf;
dbplot(X, y, @(X) sum(glmval(w_est, X, 'logit'),2));
xlabel(attributeNames(1)); ylabel(attributeNames(2));
legend(classNames);

File: Python/Scripts/ex8_2_1.py
# exercise 8.2.1


from pylab import *
from scipy.io import loadmat
from toolbox_02450 import dbplot, dbprobplot, bootstrap
from bin_classifier_ensemble import BinClassifierEnsemble
from sklearn.linear_model import LogisticRegression

# Load Matlab data file and extract variables of interest
mat_data = loadmat('../Data/synth5.mat')
X = np.matrix(mat_data['X'])
y = np.matrix(mat_data['y'])
attributeNames = [name[0] for name in mat_data['attributeNames'].squeeze()]
classNames = [name[0][0] for name in mat_data['classNames']]
N, M = X.shape
C = len(classNames)


# Fit model using bootstrap aggregation (bagging):

# Number of rounds of bagging
L = 100

# Weights for selecting samples in each bootstrap
weights = np.ones((N,1),dtype=float)/N

# Storage of trained log.reg. classifiers fitted in each bootstrap
logits = [0]*L
votes = np.zeros((N,1))

# For each round of bagging
for l in range(L):

    # Extract training set by random sampling with replacement from X and y
    X_train, y_train = bootstrap(X, y, N, weights)
    
    # Fit logistic regression model to training data and save result
    logit_classifier = LogisticRegression()
    logit_classifier.fit(X_train, y_train.A.ravel())
    logits[l] = logit_classifier
    y_est = np.mat(logit_classifier.predict(X)).T
    votes = votes + y_est

    ErrorRate = (y!=y_est).sum(dtype=float)/N
    print('Error rate: {0}%'.format(ErrorRate*100))    
    
# Estimated value of class labels (using 0.5 as threshold) by majority voting
y_est_ensemble = votes>(L/2)

# Compute error rate
ErrorRate = (y!=y_est_ensemble).sum(dtype=float)/N
print('Error rate: {:.1f}%'.format(ErrorRate*100))

ce = BinClassifierEnsemble(logits)
figure(1); dbprobplot(ce, X, y, 'auto', resolution=200)
figure(2); dbplot(ce, X, y, 'auto', resolution=200)

show()